src/shell/taskmap/hostfile.c

```diff
@@ -0,0 +1,197 @@
+/************************************************************\
+ * Copyright 2024 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* shell taskmap.hostfile plugin
+ *
+ * Read a list of hosts from a file, and assign tasks to hosts in order
+ * they are listed.
+ */
+#define FLUX_SHELL_PLUGIN_NAME "taskmap.hostfile"
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <stdio.h>
+#include <jansson.h>
+
+#include <flux/core.h>
+#include <flux/taskmap.h>
+#include <flux/hostlist.h>
+
+#include "src/common/libutil/errprintf.h"
+
+#include "builtins.h"
+
+
+/*  Create a taskmap that represents 'ntasks' tasks mapped across a set
+ *  of hosts in 'nodelist', ordered by hostlist 'hl'.
+ */
+char *taskmap_hostlist (int ntasks,
+                        struct hostlist *nodelist,
+                        struct hostlist *hl,
+                        flux_error_t *errp)
+{
+    struct taskmap *map = NULL;
+    char *result = NULL;
+    const char *host = NULL;
+
+    if (!(map = taskmap_create ()))
+        goto error;
+
+    /* Loop through hostlist hl until all tasks have been assigned to hosts
+     */
+    while (ntasks > 0) {
+        int rank;
+        if (host == NULL)
+            host = hostlist_first (hl);
+        if ((rank = hostlist_find (nodelist, host)) < 0) {
+            errprintf (errp, "host %s not found in job nodelist", host);
+            goto error;
+        }
+        if (taskmap_append (map, rank, 1, 1) < 0) {
+            errprintf (errp,
+                       "failed to append task to taskmap: %s",
+                       strerror (errno));
+            goto error;
+        }
+        host = hostlist_next (hl);
+        ntasks--;
+    }
+    result = taskmap_encode (map, TASKMAP_ENCODE_WRAPPED);
+error:
+    taskmap_destroy (map);
+    return result;
+}
+
+static struct hostlist *hostlist_from_file (const char *path)
+{
+    ssize_t n;
+    size_t len;
+    struct hostlist *hl = NULL;
+    FILE *fp;
+    char *line = NULL;
+
+    if (!(fp = fopen (path, "r")))
+        shell_die_errno (1, "failed to open hostfile: %s", path);
```

Comment:
Any particular reason why this function calls `shell_die_errno()` on error while `hostlist_from_R()` does not?

Reply:
No, that was left over from a prototype and should probably be fixed.